In [10]:
# ============================================================
# CÉLULA 1 — CONFIGURAÇÕES
# Centraliza todas as constantes do projeto
# Equivale ao arquivo de constantes/settings no Delphi
# ============================================================
import requests
import json
import unicodedata
import pandas as pd
from difflib import get_close_matches
from dataclasses import dataclass
from typing import Optional

# ----- CONSTANTES -----
SUPABASE_URL = "https://mynxlubykylncinttggu.supabase.co"
API_KEY      = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im15bnhsdWJ5a3lsbmNpbnR0Z2d1Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjUxODg2NzAsImV4cCI6MjA4MDc2NDY3MH0.Z-zqiD6_tjnF2WLU167z7jT5NzZaG72dWH0dpQW1N-Y"
IBGE_URL     = "https://servicodados.ibge.gov.br/api/v1/localidades/municipios"
SUBMIT_URL   = "https://mynxlubykylncinttggu.functions.supabase.co/ibge-submit"
FUZZY_CUTOFF = 0.85
INPUT_FILE   = "input.csv"
OUTPUT_FILE  = "resultado.csv"

HEADERS_BASE = {
    "Content-Type": "application/json",
    "apikey": API_KEY
}

import getpass

# Solicita as credenciais em tempo de execução — nunca ficam salvas no código
# Equivale a um InputBox/ShowInputQuery no Delphi
MEU_EMAIL   = input("📧 Digite seu e-mail: ")
MINHA_SENHA = getpass.getpass("🔒 Digite sua senha: ")

print("✅ [Config] Configurações carregadas!")

📧 Digite seu e-mail: andersonluizsouza25@gmail.com
🔒 Digite sua senha: ··········
✅ [Config] Configurações carregadas!


In [2]:
# ============================================================
# CÉLULA 2 — DATA CLASSES
# Define as estruturas de dados do projeto
# Equivale a Records/Structs no Delphi — tipagem explícita
# e autocompletion, evita dicionários soltos pelo código
# ============================================================

@dataclass
class Municipio:
    """Representa um município retornado pela API do IBGE"""
    nome_oficial: str
    uf:           str
    regiao:       str
    id_ibge:      int

@dataclass
class ResultadoMunicipio:
    """Representa uma linha do resultado.csv"""
    municipio_input: str
    populacao_input: int
    municipio_ibge:  str
    uf:              str
    regiao:          str
    id_ibge:         str
    status:          str  # OK | NAO_ENCONTRADO | ERRO_API

@dataclass
class Estatisticas:
    """Agrupa todas as estatísticas calculadas para envio"""
    total_municipios:     int
    total_ok:             int
    total_nao_encontrado: int
    total_erro_api:       int
    pop_total_ok:         int
    medias_por_regiao:    dict

    def to_dict(self) -> dict:
        """Serializa para o formato JSON esperado pela API
        Equivale a um método ToJSON numa classe Delphi"""
        return {
            "total_municipios":     self.total_municipios,
            "total_ok":             self.total_ok,
            "total_nao_encontrado": self.total_nao_encontrado,
            "total_erro_api":       self.total_erro_api,
            "pop_total_ok":         self.pop_total_ok,
            "medias_por_regiao":    self.medias_por_regiao
        }

print("✅ [DataClasses] Estruturas de dados definidas!")

✅ [DataClasses] Estruturas de dados definidas!


In [3]:
# ============================================================
# CÉLULA 3 — AUTH SERVICE
# Responsabilidade única: autenticação no Supabase
# Equivale a uma classe TSupabaseAuth no Delphi
# Padrão: exceptions tipadas para cada tipo de falha
# ============================================================

class AuthError(Exception):
    """Erro específico de autenticação — mais descritivo que Exception genérica.
    Em Delphi seria um EAuthError herdando de Exception"""
    pass

class AuthService:
    """Gerencia o ciclo de autenticação com o Supabase.
    Padrão Service Layer — isola toda lógica de auth num único lugar"""

    def __init__(self, supabase_url: str, headers_base: dict):
        self.supabase_url   = supabase_url
        self.headers_base   = headers_base
        self.access_token:  Optional[str] = None  # None = não autenticado

    def login(self, email: str, senha: str) -> str:
        """Autentica e armazena o token internamente.
        Lança AuthError em caso de falha — nunca retorna silenciosamente"""
        url = f"{self.supabase_url}/auth/v1/token?grant_type=password"

        try:
            response = requests.post(
                url,
                headers=self.headers_base,
                json={"email": email, "password": senha},
                timeout=10  # Evita travar indefinidamente — boa prática de produção
            )
        except requests.exceptions.Timeout:
            raise AuthError("Timeout ao conectar com o Supabase")
        except requests.exceptions.ConnectionError:
            raise AuthError("Sem conexão com o Supabase")

        if response.status_code != 200:
            raise AuthError(
                f"Login falhou [HTTP {response.status_code}]: {response.text}"
            )

        self.access_token = response.json()["access_token"]
        return self.access_token

    def get_auth_headers(self) -> dict:
        """Retorna headers com Bearer Token para requisições autenticadas.
        Garante que login foi feito antes de usar — fail fast"""
        if not self.access_token:
            raise AuthError("Não autenticado. Chame login() primeiro.")
        return {
            "Content-Type":  "application/json",
            "Authorization": f"Bearer {self.access_token}"
        }

# Instanciando e testando
auth_service = AuthService(SUPABASE_URL, HEADERS_BASE)
auth_service.login(MEU_EMAIL, MINHA_SENHA)
print("✅ [AuthService] Login realizado com sucesso!")

✅ [AuthService] Login realizado com sucesso!


In [4]:
# ============================================================
# CÉLULA 4 — IBGE SERVICE
# Responsabilidade única: buscar e indexar municípios do IBGE
# Equivale a uma classe TIBGEService no Delphi
# Padrão: cache em memória — 1 requisição para ~5500 municípios
# ============================================================

class IBGEError(Exception):
    """Erro específico de integração com o IBGE"""
    pass

class IBGEService:
    """Busca municípios do IBGE e oferece lookup normalizado.
    Estratégia: carrega tudo em memória uma vez — O(1) nas buscas"""

    def __init__(self, ibge_url: str):
        self.ibge_url  = ibge_url
        self._index:   dict = {}  # Dicionário interno — como TDictionary no Delphi
        self._carregado = False

    def _normalizar(self, texto: str) -> str:
        """Remove acentos e padroniza case para comparação.
        Privado (prefixo _) — detalhe de implementação, não API pública.
        Equivale a um método private no Delphi"""
        texto = texto.lower().strip()
        texto = unicodedata.normalize('NFD', texto)
        return ''.join(c for c in texto if unicodedata.category(c) != 'Mn')

    def _extrair_uf_regiao(self, municipio: dict) -> tuple:
        """Extrai UF e região navegando no JSON aninhado do IBGE.
        Trata nulos com fallback — alguns municípios têm dados incompletos"""
        try:
            uf     = municipio["microrregiao"]["mesorregiao"]["UF"]["sigla"]
            regiao = municipio["microrregiao"]["mesorregiao"]["UF"]["regiao"]["nome"]
            return uf, regiao
        except (TypeError, KeyError):
            pass

        try:
            uf     = municipio["regiao-imediata"]["regiao-intermediaria"]["UF"]["sigla"]
            regiao = municipio["regiao-imediata"]["regiao-intermediaria"]["UF"]["regiao"]["nome"]
            return uf, regiao
        except (TypeError, KeyError):
            return "N/A", "N/A"

    def carregar(self) -> None:
        """Baixa e indexa todos os municípios do IBGE em memória.
        Separado do __init__ para controle explícito do momento da carga"""
        print("⏳ [IBGEService] Buscando municípios...")

        try:
            response = requests.get(self.ibge_url, timeout=30)
        except requests.exceptions.Timeout:
            raise IBGEError("Timeout ao conectar com a API do IBGE")
        except requests.exceptions.ConnectionError:
            raise IBGEError("Sem conexão com a API do IBGE")

        if response.status_code != 200:
            raise IBGEError(f"API IBGE retornou HTTP {response.status_code}")

        for m in response.json():
            uf, regiao = self._extrair_uf_regiao(m)
            chave = self._normalizar(m["nome"])
            self._index[chave] = Municipio(
                nome_oficial = m["nome"],
                uf           = uf,
                regiao       = regiao,
                id_ibge      = m["id"]
            )

        self._carregado = True
        print(f"✅ [IBGEService] {len(self._index)} municípios indexados!")

    def buscar(self, nome: str) -> Optional[Municipio]:
        """Busca um município por nome com fallback fuzzy.
        Retorna None se não encontrar — nunca lança exceção para não encontrado"""
        if not self._carregado:
            raise IBGEError("Índice não carregado. Chame carregar() primeiro.")

        chave = self._normalizar(nome)

        # Tentativa 1: match exato — O(1)
        if chave in self._index:
            return self._index[chave]

        # Tentativa 2: fuzzy matching — corrige erros de digitação
        matches = get_close_matches(
            chave,
            self._index.keys(),
            n=1,
            cutoff=FUZZY_CUTOFF
        )
        if matches:
            print(f"  🔧 [IBGEService] Fuzzy: '{nome}' → '{self._index[matches[0]].nome_oficial}'")
            return self._index[matches[0]]

        return None  # NAO_ENCONTRADO — tratado pela camada acima

# Instanciando e carregando
ibge_service = IBGEService(IBGE_URL)
ibge_service.carregar()

⏳ [IBGEService] Buscando municípios...
✅ [IBGEService] 5290 municípios indexados!


In [5]:
# ============================================================
# CÉLULA 5 — PROCESSOR
# Responsabilidade única: orquestrar o matching e gerar resultados
# Equivale a uma classe TProcessor no Delphi
# Padrão: recebe dependências prontas (Injeção de Dependência)
# ============================================================

class Processor:
    """Processa o CSV e aplica as regras de negócio de matching.
    Recebe IBGEService como dependência — facilita testes e substituição"""

    def __init__(self, ibge_service: IBGEService):
        self.ibge_service = ibge_service

    def _criar_resultado_ok(
        self,
        municipio_input: str,
        populacao_input: int,
        municipio: Municipio
    ) -> ResultadoMunicipio:
        """Factory method para resultado OK — evita repetição de código.
        Equivale a um constructor alternativo no Delphi"""
        return ResultadoMunicipio(
            municipio_input = municipio_input,
            populacao_input = populacao_input,
            municipio_ibge  = municipio.nome_oficial,
            uf              = municipio.uf,
            regiao          = municipio.regiao,
            id_ibge         = str(municipio.id_ibge),
            status          = "OK"
        )

    def _criar_resultado_nao_encontrado(
        self,
        municipio_input: str,
        populacao_input: int
    ) -> ResultadoMunicipio:
        """Factory method para resultado NAO_ENCONTRADO"""
        return ResultadoMunicipio(
            municipio_input = municipio_input,
            populacao_input = populacao_input,
            municipio_ibge  = "",
            uf              = "",
            regiao          = "",
            id_ibge         = "",
            status          = "NAO_ENCONTRADO"
        )

    def processar(self, df: pd.DataFrame) -> list[ResultadoMunicipio]:
        """Processa cada linha do DataFrame aplicando as regras de negócio:
        1. Busca exata no índice IBGE
        2. Fallback fuzzy para erros de digitação
        3. Deduplicação — duplicatas corrompidas viram NAO_ENCONTRADO
        Equivale a um método Execute numa TThread do Delphi"""

        resultados = []
        # Set de controle de duplicatas — como TStringList no Delphi
        municipios_processados: set = set()

        for _, row in df.iterrows():
            municipio_input = row["municipio"]
            populacao_input = int(row["populacao"])

            municipio = self.ibge_service.buscar(municipio_input)

            if municipio is None:
                # Não encontrado nem com fuzzy — erro de digitação irrecuperável
                resultados.append(
                    self._criar_resultado_nao_encontrado(municipio_input, populacao_input)
                )
                continue

            if municipio.nome_oficial in municipios_processados:
                # Duplicata detectada — mesmo município oficial já foi processado
                print(f"  ⚠️  [Processor] Duplicata: '{municipio_input}' → '{municipio.nome_oficial}'")
                resultados.append(
                    self._criar_resultado_nao_encontrado(municipio_input, populacao_input)
                )
                continue

            municipios_processados.add(municipio.nome_oficial)
            resultados.append(
                self._criar_resultado_ok(municipio_input, populacao_input, municipio)
            )

        return resultados

# Instanciando e testando
processor = Processor(ibge_service)
print("✅ [Processor] Pronto!")

✅ [Processor] Pronto!


In [6]:
# ============================================================
# CÉLULA 6 — STATS CALCULATOR + EXPORTER
# Responsabilidade: calcular estatísticas e exportar CSV
# Duas classes pequenas e coesas — cada uma faz uma coisa só
# Equivale a TStatsCalculator e TFileExporter no Delphi
# ============================================================

class StatsCalculator:
    """Calcula todas as estatísticas a partir da lista de resultados.
    Recebe dados puros — sem dependência de arquivo ou API"""

    def calcular(self, resultados: list[ResultadoMunicipio]) -> Estatisticas:
        """Calcula e retorna um objeto Estatisticas tipado.
        Usar DataFrame só aqui — isola a dependência do Pandas"""

        df = pd.DataFrame([vars(r) for r in resultados])
        # vars() converte dataclass para dict — como record.AsJSON no Delphi

        df_ok = df[df["status"] == "OK"]

        # Média por região — GROUP BY regiao + AVG(populacao) em SQL
        medias = (
            df_ok.groupby("regiao")["populacao_input"]
            .mean()
            .round(2)
            .to_dict()
        )

        return Estatisticas(
            total_municipios     = len(df),
            total_ok             = len(df_ok),
            total_nao_encontrado = len(df[df["status"] == "NAO_ENCONTRADO"]),
            total_erro_api       = len(df[df["status"] == "ERRO_API"]),
            pop_total_ok         = int(df_ok["populacao_input"].sum()),
            medias_por_regiao    = medias
        )


class Exporter:
    """Exporta os resultados para CSV.
    Separado do StatsCalculator — responsabilidade única"""

    def exportar(self, resultados: list[ResultadoMunicipio], caminho: str) -> None:
        """Salva a lista de resultados como CSV.
        Equivale ao FDMemTable.SaveToFile() no Delphi"""
        df = pd.DataFrame([vars(r) for r in resultados])
        df.to_csv(caminho, index=False, encoding='utf-8')
        print(f"✅ [Exporter] '{caminho}' gerado com {len(df)} linhas!")


# Instanciando
stats_calculator = StatsCalculator()
exporter         = Exporter()
print("✅ [StatsCalculator] Pronto!")
print("✅ [Exporter] Pronto!")

✅ [StatsCalculator] Pronto!
✅ [Exporter] Pronto!


In [7]:
# ============================================================
# CÉLULA 7 — SUBMITTER
# Responsabilidade única: enviar estatísticas para a API de correção
# Equivale a uma classe TAPISubmitter no Delphi
# Padrão: recebe AuthService como dependência — nunca manipula token diretamente
# ============================================================

class SubmitError(Exception):
    """Erro específico de envio para a API de correção"""
    pass

class Submitter:
    """Envia as estatísticas calculadas para a API de correção da Nasajon.
    Depende do AuthService para obter o Bearer Token — nunca acessa token diretamente"""

    def __init__(self, submit_url: str, auth_service: AuthService):
        self.submit_url   = submit_url
        self.auth_service = auth_service  # Injeção de dependência

    def enviar(self, stats: Estatisticas) -> dict:
        """Monta o payload, envia e retorna a resposta completa da API.
        Lança SubmitError em caso de falha — nunca falha silenciosamente"""

        payload = {"stats": stats.to_dict()}

        print("⏳ [Submitter] Enviando para a API de correção...")
        print(f"   Payload: {json.dumps(payload, indent=2, ensure_ascii=False)}")

        try:
            response = requests.post(
                self.submit_url,
                headers = self.auth_service.get_auth_headers(),
                json    = payload,
                timeout = 15
            )
        except requests.exceptions.Timeout:
            raise SubmitError("Timeout ao conectar com a API de correção")
        except requests.exceptions.ConnectionError:
            raise SubmitError("Sem conexão com a API de correção")

        if response.status_code != 200:
            raise SubmitError(
                f"API retornou HTTP {response.status_code}: {response.text}"
            )

        resultado = response.json()
        print(f"\n🏆 [Submitter] Score: {resultado['score']}")
        print(f"💬 [Submitter] Feedback: {resultado['feedback']}")
        return resultado

# Instanciando
submitter = Submitter(SUBMIT_URL, auth_service)
print("✅ [Submitter] Pronto!")

✅ [Submitter] Pronto!


In [11]:
# ============================================================
# CÉLULA 8 — MAIN
# Orquestra todas as etapas em sequência
# Equivale ao blanco begin..end do .dpr no Delphi
# Padrão: try/except no nível mais alto — captura qualquer falha
# ============================================================

def criar_input_csv() -> None:
    """Cria o input.csv com os dados do desafio"""
    conteudo = """municipio,populacao
Niteroi,515317
Sao Gonçalo,1091737
Sao Paulo,12396372
Belo Horzionte,2530701
Florianopolis,516524
Santo Andre,723889
Santoo Andre,700000
Rio de Janeiro,6718903
Curitba,1963726
Brasilia,3094325"""

    with open(INPUT_FILE, 'w', encoding='utf-8') as f:
        f.write(conteudo)
    print(f"✅ [Main] '{INPUT_FILE}' criado!")

def main():
    print("=" * 55)
    print("   DESAFIO TÉCNICO NASAJON — Anderson Souza")
    print("=" * 55)

    try:
        # ETAPA 0: Criar input.csv
        criar_input_csv()

        # ETAPA 1: Autenticação
        print("\n🔐 Etapa 1 — Autenticação...")
        auth_service.login(MEU_EMAIL, MINHA_SENHA)

        # ETAPA 2: Leitura do CSV
        print("\n📋 Etapa 2 — Leitura do CSV...")
        df = pd.read_csv(INPUT_FILE)
        print(f"   {len(df)} registros carregados!")

        # ETAPA 3: Carga do IBGE
        print("\n🗺️  Etapa 3 — Carga da API do IBGE...")
        ibge_service.carregar()

        # ETAPA 4: Processamento e matching
        print("\n🔍 Etapa 4 — Matching dos municípios...")
        resultados = processor.processar(df)

        # ETAPA 5: Exportação do resultado.csv
        print("\n💾 Etapa 5 — Exportando resultado.csv...")
        exporter.exportar(resultados, OUTPUT_FILE)

        # Exibindo resultado
        df_resultado = pd.DataFrame([vars(r) for r in resultados])
        print(df_resultado.to_string())

        # ETAPA 6: Cálculo de estatísticas
        print("\n📊 Etapa 6 — Calculando estatísticas...")
        stats = stats_calculator.calcular(resultados)
        print(f"   total_municipios:     {stats.total_municipios}")
        print(f"   total_ok:             {stats.total_ok}")
        print(f"   total_nao_encontrado: {stats.total_nao_encontrado}")
        print(f"   total_erro_api:       {stats.total_erro_api}")
        print(f"   pop_total_ok:         {stats.pop_total_ok}")
        print(f"   medias_por_regiao:    {stats.medias_por_regiao}")

        # ETAPA 7: Envio final
        print("\n📤 Etapa 7 — Enviando para a API de correção...")
        resultado_final = submitter.enviar(stats)

        print("\n" + "=" * 55)
        print(f"   🏆 SCORE FINAL: {resultado_final['score']}/10")
        print(f"   💬 {resultado_final['feedback']}")
        print("=" * 55)

    except AuthError as e:
        print(f"\n❌ [Auth] {e}")
    except IBGEError as e:
        print(f"\n❌ [IBGE] {e}")
    except SubmitError as e:
        print(f"\n❌ [Submit] {e}")
    except Exception as e:
        print(f"\n❌ [Erro inesperado] {e}")

# Executando
main()

   DESAFIO TÉCNICO NASAJON — Anderson Souza
✅ [Main] 'input.csv' criado!

🔐 Etapa 1 — Autenticação...

📋 Etapa 2 — Leitura do CSV...
   10 registros carregados!

🗺️  Etapa 3 — Carga da API do IBGE...
⏳ [IBGEService] Buscando municípios...
✅ [IBGEService] 5290 municípios indexados!

🔍 Etapa 4 — Matching dos municípios...
  🔧 [IBGEService] Fuzzy: 'Belo Horzionte' → 'Belo Horizonte'
  🔧 [IBGEService] Fuzzy: 'Santoo Andre' → 'Santo André'
  ⚠️  [Processor] Duplicata: 'Santoo Andre' → 'Santo André'
  🔧 [IBGEService] Fuzzy: 'Curitba' → 'Curitiba'

💾 Etapa 5 — Exportando resultado.csv...
✅ [Exporter] 'resultado.csv' gerado com 10 linhas!
  municipio_input  populacao_input  municipio_ibge  uf        regiao  id_ibge          status
0         Niteroi           515317         Niterói  RJ       Sudeste  3303302              OK
1     Sao Gonçalo          1091737     São Gonçalo  RJ       Sudeste  3304904              OK
2       Sao Paulo         12396372       São Paulo  SP       Sudeste  3550308  

# 📝 Notas Explicativas — Decisões Técnicas

## Linguagem e Ambiente
Python no Google Colab — escolhido pela agilidade no tratamento de dados
e por não exigir instalação de ambiente local.

## Arquitetura — Service Layer
O código foi organizado em camadas com responsabilidade única:
- **AuthService** — autenticação e gestão do token JWT
- **IBGEService** — integração e cache dos municípios do IBGE
- **Processor** — regras de negócio do matching
- **StatsCalculator** — cálculo das estatísticas
- **Exporter** — geração do resultado.csv
- **Submitter** — envio para a API de correção

## Estratégia de Matching
1. **Match exato normalizado** — remove acentos e diferenças de case
2. **Fuzzy matching** (difflib, cutoff=0.85) — corrige erros de digitação
   como `Belo Horzionte` → `Belo Horizonte` e `Curitba` → `Curitiba`
3. **Deduplicação** — `Santoo Andre` é detectado como duplicata corrompida
   de `Santo Andre` e marcado como `NAO_ENCONTRADO`

## Performance
Uma única requisição ao IBGE (~5290 municípios) indexada em dicionário
em memória — busca O(1) ao invés de N requisições individuais.

## Tratamento de Erros
Exceptions tipadas por camada (AuthError, IBGEError, SubmitError)
com timeout em todas as requisições HTTP.

## O que faria diferente em produção
- Variáveis sensíveis (email, senha) em variáveis de ambiente (.env)
- Cache persistente do índice IBGE para evitar recarga a cada execução
- Testes unitários para cada Service
- Logging estruturado ao invés de print()